util/tracer_nvbit/tracer_tool/tracer_tool.cu

```diff
@@ -348,182 +549,152 @@ void nvbit_at_cuda_event(CUcontext ctx, int is_exit, nvbit_api_cuda_t cbid,
         }
-
-        fprintf(ctx_resultsFile[ctx], "-kernel name = %s\n",
-                nvbit_get_func_name(ctx, p->f, true));
-        fprintf(ctx_resultsFile[ctx], "-kernel id = %d\n", ctx_kernelid[ctx]);
-        fprintf(ctx_resultsFile[ctx], "-grid dim = (%d,%d,%d)\n", p->gridDimX,
-                p->gridDimY, p->gridDimZ);
-        fprintf(ctx_resultsFile[ctx], "-block dim = (%d,%d,%d)\n", p->blockDimX,
-                p->blockDimY, p->blockDimZ);
-        fprintf(ctx_resultsFile[ctx], "-shmem = %d\n",
-                shmem_static_nbytes + p->sharedMemBytes);
-        fprintf(ctx_resultsFile[ctx], "-nregs = %d\n", nregs);
-        fprintf(ctx_resultsFile[ctx], "-binary version = %d\n", binary_version);
-        fprintf(ctx_resultsFile[ctx], "-cuda stream id = %lu\n", (uint64_t)p->hStream);
-        fprintf(ctx_resultsFile[ctx], "-shmem base_addr = 0x%016lx\n",
-                (uint64_t)nvbit_get_shmem_base_addr(ctx));
-        fprintf(ctx_resultsFile[ctx], "-local mem base_addr = 0x%016lx\n",
-                (uint64_t)nvbit_get_local_mem_base_addr(ctx));
-        fprintf(ctx_resultsFile[ctx], "-nvbit version = %s\n", NVBIT_VERSION);
-        fprintf(ctx_resultsFile[ctx], "-accelsim tracer version = %s\n", TRACER_VERSION);
-        fprintf(ctx_resultsFile[ctx], "-enable lineinfo = %d\n", lineinfo);
-        fprintf(ctx_resultsFile[ctx], "\n");
-
-        fprintf(ctx_resultsFile[ctx],
-                "#traces format = [line_num] PC mask dest_num [reg_dests] "
-                "opcode src_num "
-                "[reg_srcs] mem_width [adrrescompress?] [mem_addresses] "
-                "immediate\n");
-        fprintf(ctx_resultsFile[ctx], "\n");
       }
+    }
+  } break;
+  case API_CUDA_cuGraphAddKernelNode: {
+    cuGraphAddKernelNode_params *p = (cuGraphAddKernelNode_params *)params;
+    CUfunction func = p->nodeParams->func;
 
-      kernelsFile = fopen(ctx_kernelslist[ctx].c_str(), "a");
-      // This will be a relative path to the traces file
-      
-      sprintf(buffer, "kernel-%d-ctx_0x%lx.trace%s", ctx_kernelid[ctx], ctx,
-              xz_compress_trace ? ".xz" : "");
-      if (!stop_report) {
-        fprintf(kernelsFile, buffer);
-        fprintf(kernelsFile, "\n");
-      }
-      fclose(kernelsFile);
-
-      statsFile = fopen(ctx_stats_location[ctx].c_str(), "a");
-      unsigned blocks = p->gridDimX * p->gridDimY * p->gridDimZ;
-      unsigned threads = p->blockDimX * p->blockDimY * p->blockDimZ;
-
-      fprintf(statsFile, "%s, %s, %d, %d, %d, %d, %d, %d, %d, %d, ", buffer,
-              nvbit_get_func_name(ctx, p->f, true), p->gridDimX, p->gridDimY,
-              p->gridDimZ, blocks, p->blockDimX, p->blockDimY, p->blockDimZ,
-              threads);
-
-      fclose(statsFile);
-
-      ctx_kernelid[ctx]++;
-      recv_thread_receiving = true;
-
-    } else {
-      /* make sure current kernel is completed */
-      cudaDeviceSynchronize();
+    if (!is_exit) {
```

Comment:
Is every CUDA graph node, by definition, a Kernel node or an exit node?

Reply:
I am not sure I follow this one. This API here just builds the graph manually by adding 1 node.

From my understanding, the instrumentation of CUDA Graph is done at the graph-building phase. However, since a graph might not launch at the moment of building, we have to wait till it is actually launched and completed to call the `leave_kernel_launch()`, which is at the `API_CUDA_cuGraphLaunch`